### DISCUSSION TITLE
Firmware Retract

### DISCUSSION THREAD

**Discussion by Eifel-Joe** [OrcaSlicer/OrcaSlicer#70] [Ideas]
Hello SoftFever,

is it possible to add the posibility of choosing wheather retract controlled by slicer or by firmware (G10/G11)?

I just came from SuperSlicer in combination with klipper driven printers. All my printers are configured to use firmware retract so that i can change retract settings during print.

It would be another great advantage for your slicer. Not just fpr me because i know lots of people using it.

**Comment by SoftFever** [comment]
Yeah, this is a good idea

**Reply by boomer20021** [reply]
Is there a way to add a retraction test in to I think that will fix the string to the next update. softfever

**Comment by Tyler8299** [comment]
I as well would like to see this. thank you for your work

**Comment by SoftFever** [comment]
This feature has been added.
Will be available in the coming update.

**Comment by Tyler8299** [comment]
You are amazing thank you
________________________________

**Comment by MicroSur** [comment]
But how to set the initial parameters for FW retraction?
I redefined the speed in the filament parameters, but the changes were not applied...

**Comment by dabstractor** [comment]
I'm confused by this too. I'm setting my retraction settings in orcaslicer, I have "use firmware retraction" enabled, but klipper is still using the retraction settings from printer.cfg. The command for setting the retraction value (SET_RETRACTION) is not being added to my gcode file. Am I missing something? Am I supposed to wire this up by hand? I thought the slicer handled it.

**Comment by virajmalia** [comment]
@MicroSur @dabstractor 
Firmware(FW) retraction means that your klipper/marlin/custom firmware will decide `how much and how fast` to retract, while the gcode from the slicer will only tell the FW `when` to retract during the printing process.
This works as expected because when I read the retraction tower calibration print gcode, it shows the G10/G11 commands.

> But how to set the initial parameters for FW retraction?

https://www.klipper3d.org/Config_Reference.html#firmware_retraction This section shows the config parameters that should be set in your printer.cfg. These will be used when you enable FW retraction in your slicer.

### Calibration

#### FW retraction disabled
Orca Slicer's retraction tower works out of the box.

#### FW retraction enabled
You can change the retraction settings `during the print` from your Fluid or Mainsail WebUI.

Orca slicer's retraction tower test cannot change retraction settings on layer change automatically. I wonder if I need to add some gcode to the layer change section.
I have the option enabled already:
![image](https://github.com/user-attachments/assets/dee53b4c-b4e5-47c2-8aa1-443bc5fc942f)

Note that this only affects the dynamic changing of retraction settings. For all purposes, if you want to use the fixed retraction settings in your printer.cfg, then this is not a problem - although it becomes challenging to perform calibration.

One workaround I found:
1. Disable FW retraction and perform calibration using Orca slicer's retraction tower. Note down the 3 values: retract length, speed and unretract speed.
2. Enable FW retraction and use the above noted values in your printer.cfg.

**Comment by MicroSur** [comment]
@virajmalia 
Editing the config every time I change filament - is a bad idea. The slicer should take my overridden parameters from the filament config section and set SET_RETRACTION with the changes in the g-code. But no.

I have to put SET_RETRACTION line in the filament start g-code field.

**Comment by R0cketRacc00n** [comment]
I wrote a universal macro for overriding klipper retraction settings in the Orca and Prusa slicers.
```
; Filament gcode
; Autor Fayrzakhmanov Azat
{if not (is_nil(filament_retraction_length[filament_extruder_id]) && is_nil(filament_retraction_speed[filament_extruder_id]) && is_nil(filament_deretraction_speed[filament_extruder_id]))
    }SET_RETRACTION{
    if not is_nil(filament_retraction_length[filament_extruder_id])
        } RETRACT_LENGTH={filament_retraction_length[filament_extruder_id]}{
    endif
    if not is_nil(filament_retraction_speed[filament_extruder_id])
        } RETRACT_SPEED={filament_retraction_speed[filament_extruder_id]}{
    endif
    if not is_nil(filament_deretraction_speed[filament_extruder_id])
        } UNRETRACT_SPEED={filament_deretraction_speed[filament_extruder_id]}{
    endif
    if not is_nil(filament_retract_restart_extra[filament_extruder_id])
        } UNRETRACT_EXTRA_LENGTH={filament_retract_restart_extra[filament_extruder_id]}{
    endif}{
endif}
```
